src/tools/tidy/src/ext_tool_checks.rs

```diff
@@ -0,0 +1,368 @@
+            // Skip insufficient version errors
+            Err(Error::Version { installed, .. }) => found.push(installed),
+            // just log and skip unrecognized errors
+            Err(e) => eprintln!("note: error running '{py}': {e}"),
+        }
+    }
+
+    let Some(sys_py) = sys_py else {
+        let ret = if found.is_empty() {
+            Error::MissingReq("python3", "python file checks", None)
+        } else{
+            found.sort();
+            found.dedup();
+            Error::Version { program: "python3", required: MIN_PY_REV_STR, installed: found.join(", ") }
+        };
+        return Err(ret);
+    };
+
+    eprintln!("creating virtual environment at '{}' using '{sys_py}'", path.display());
+    let out = Command::new(sys_py).args(["-m", "virtualenv"]).arg(path).output().unwrap();
+
+    if out.status.success() {
+        Ok(())
+    } else if String::from_utf8_lossy(&out.stderr).contains("No module named virtualenv") {
+        Err(Error::Generic(format!("virtualenv not found: is it installed for {sys_py}?")))
+    } else {
+        Err(Error::Generic(format!(
+            "failed to create virtualenv at '{}' using {sys_py}",
+            path.display()
+        )))
+    }
+}
+
+/// Parse python's version output (`Python x.y.z`) and ensure we have a
+/// suitable version.
+fn verify_py_version(py_path: &Path) -> Result<(), Error> {
+    let out = Command::new(py_path).arg("--version").output()?;
+    let outstr = String::from_utf8_lossy(&out.stdout);
+    let vers = outstr.trim().split_ascii_whitespace().nth(1).unwrap().trim();
+    let mut vers_comps = vers.split('.');
+    let major: u32 = vers_comps.next().unwrap().parse().unwrap();
+    let minor: u32 = vers_comps.next().unwrap().parse().unwrap();
+
+    if (major < MIN_PY_REV.0) || ((major == MIN_PY_REV.0) && (minor < MIN_PY_REV.1)) {
+        Err(Error::Version {
+            program: "python",
+            required: MIN_PY_REV_STR,
+            installed: vers.to_owned(),
+        })
+    } else {
+        Ok(())
+    }
+}
+
+/// Given a binary (e.g. `black`, `ruff`) initialize it in the venv and return its path
+fn get_or_init_py_bin(
+    py_path: &Path,
+    bin: &str,
+    version: &str,
+    pip_checked: &mut bool,
+) -> Result<(), Error> {
+    let bin_vers = format!("{bin}{version}");
+    let out = Command::new(py_path).args(["-m", "pip", "freeze"]).output().unwrap().stdout;
+
+    if String::from_utf8_lossy(&out).contains(&bin_vers) {
+        return Ok(());
+    }
+
+    eprintln!("installing {bin_vers} via pip");
+
+    if !*pip_checked {
+        // verify pip is updated before installing anything
+        Command::new(py_path)
+            .args(["-m", "pip", "install", "--upgrade", "pip"])
+            .status()
+            .expect("failed to launch python");
+        *pip_checked = true;
+    }
+
+    let status = Command::new(py_path).args(["-m", "pip", "install", &bin_vers]).status()?;
```

Comment:
Installing any version could lead to differences between users. What I recommend is:

* Add a `requirements.txt` somewhere in the source tree, pinning the versions of the tools we use (ideally with `pip-compile --generate-hashes` from a `requirements.in`)
* When creating a virtualenv, install the `requirements.txt` file and copy the file itself into the virtualenv
* In any future tidy invocation, check if the `requirements.txt` in the source tree matches the one copied into the virtualenv, otherwise discard the whole virtualenv and create a new one from scratch.

Reply:
I do like your process and will update to do that so versions aren't hard coded. But what differences are possible as-is? `bin_vers` is a pinned version e.g. `ruff==0.0.272`, and the version is verified before running https://github.com/rust-lang/rust/blob/1d05b0dbb4ce2e0afa82d63518443ba96aaf7888/src/tools/tidy/src/ext_tool_checks.rs#L248-L251

Reply:
Before we do a bunch of work with pinning requirements, I think it's useful to take a step back and ask whether we want this checking at all. I'm not convinced I want to put contributors on the hook for fixing lints and formatting in the majority of our Python and shell code. If it's not checked in CI then I'm not sure it adds enough value for us to then take on the burden of maintaining this code and getting PRs trying to fix issues these linters will warn about.